src/ImpactEvaluator.sol

```diff
@@ -134,18 +136,49 @@ contract ImpactEvaluator is AccessControl {
         uint64[] memory scores
     ) private {
         for (uint i = 0; i < addresses.length; i++) {
-            address payable addr = addresses[i];
-            if (addr == 0x000000000000000000000000000000000000dEaD) {
-                // TODO: Shall we emit an event here too?
-                continue;
-            }
-            uint256 amount = (scores[i] * roundReward) / MAX_SCORE;
-            if (addr.send(amount)) {
-                emit Transfer(addr, amount);
-            } else {
-                emit TransferFailed(addr, amount);
-            }
+            balances[addresses[i]] += (scores[i] * roundReward) / MAX_SCORE;
```

Comment:
What happens when we run out of funds? IIUC, we will still increase the balance of participants. Then we have a bank run - the first participant to withdraw their rewards will get them, others will not.

I think it's crucial to verify that we have enough *remaining balance* before assigning rewards to participants. After we assign the rewards, the *remaining balance* must be adjusted accordingly.

(There are different ways how to implement this. We can keep track of the *remaining balance* as another state field, but we can also calculate it on the fly by summing up `balances`.)

IMO, we should also restore the feature that `0x0...0dEAD` does not receive any rewards and therefore does not change the *remaining balance*.

Reply:
moved to Slack https://filecoinproject.slack.com/archives/C03PQG6UT2B/p1698924505273569